src/scenic/core/scenarios.py

```diff
@@ -350,77 +404,44 @@ def generate(self, maxIterations=2000, verbosity=0, feedback=None):
-
-        for instance in self._instances:
-            vi = sample[instance]
-
-            # Require that the observing entity, if one has been specified,
-            # can see the instance.
-            if vi._observingEntity is not None:
-                observing_entity = sample[vi._observingEntity]
-                if not observing_entity.canSee(vi, occludingObjects=occluders(observing_entity, vi)):
-                    return 'instance visibility (from observing entity)'
-
-            # Require that the non-observing entity, if one has been specified,
-            # can see the instance.
-            if vi._nonObservingEntity is not None:
-                non_observing_entity = sample[vi._nonObservingEntity]
-                if non_observing_entity.canSee(vi, occludingObjects=occluders(non_observing_entity, vi)):
-                    return 'instance visibility (from non-observing entity)'
-
-        # Check built-in requirements for objects
-        for i, vi in enumerate(sampledObjects):
-            # Require object to be contained in the workspace/valid region
-            container = self.containerOfObject(vi)
-            if not container.containsObject(vi):
-                return 'object containment'
+        ## Optional Requirements ##
+        # Any collision indicates an intersection
+        requirements.append(BlanketCollisionRequirement(self.objects))
 
-            # Require object to be visible from the ego object
-            if vi.requireVisible and vi is not ego:
-                if not ego.canSee(vi, occludingObjects=occluders(ego, vi)):
-                    return 'object visibility (from ego)'
-
-            # Require object to not intersect another object
-            if not vi.allowCollisions:
-                for vj in sampledObjects[(i+1):]:
-                    if not vj.allowCollisions and vi.intersects(vj):
-                        return 'object intersection'
-
-        # Check user-specified requirements
-        for req in activeReqs:
-            if req.falsifiedBy(sample):
-                return str(req)
-
-        # No rejection, return None
-        return None
+        ## Mandatory Requirements ##
+        # Pairwise object intersection
+        colliding_objects = (obj for obj in self.objects if 
+            needsSampling(obj.allowCollisions) or not obj.allowCollisions)
+        for objA, objB in itertools.combinations(colliding_objects, 2):
+            requirements.append(IntersectionRequirement(objA, objB))
+
+        # Object containment
+        for obj in self.objects:
+            container = self.containerOfObject(obj)
+            if not isinstance(container, AllRegion):
+                requirements.append(ContainmentRequirement(obj, container))
+
+        # Observing entity visibility
+        for obj in filter(lambda x: x._observingEntity is not None, self._instances):
+            requirements.append(VisibilityRequirement(obj._observingEntity,
+                obj, self.objects))
```

Comment:
Might as well save time later by filtering out known non-occluding objects here, like you do with `colliding_objects` above.

Reply:
Good call.